conda_lock/src_parser/__init__.py

```diff
@@ -1,39 +1,207 @@
 import hashlib
 import json
 
-from typing import List, Optional
+from collections import defaultdict
+from dataclasses import dataclass
+from itertools import chain
+from typing import Dict, List, Literal, Optional, Sequence, Set, Tuple, TypedDict
+
+from pydantic import BaseModel, Field
 
 from conda_lock.virtual_package import FakeRepoData
 
 
-class LockSpecification:
-    def __init__(
-        self,
-        specs: List[str],
-        channels: List[str],
-        platform: str,
-        virtual_package_repo: Optional[FakeRepoData] = None,
-    ):
-        self.specs = specs
-        self.channels = channels
-        self.platform = platform
-        self.virtual_package_repo = virtual_package_repo
+class StrictModel(BaseModel):
+    class Config:
+        extra = "forbid"
+
+
+class Selectors(StrictModel):
+    platform: Optional[List[str]] = None
+
+    def __ior__(self, other) -> "Selectors":
+        if not isinstance(other, Selectors):
+            raise TypeError
+        if other.platform and self.platform:
+            for p in other.platform:
+                if p not in self.platform:
+                    self.platform.append(p)
+        return self
+
+
+class Dependency(StrictModel):
+    name: str
+    manager: Literal["conda", "pip"] = "conda"
+    optional: bool = False
+    category: str = "main"
+    extras: List[str] = []
+    selectors: Selectors = Selectors()
+
+
+class VersionedDependency(Dependency):
+    version: str
+
+
+class URLDependency(Dependency):
+    url: str
+    hashes: List[str]
+
+
+class Package(StrictModel):
+    url: str
+    hash: str
+
+
+class DependencySource(StrictModel):
+    type: Literal["url"]
+    url: str
+
+
+class LockedDependency(StrictModel):
+    name: str
+    version: str
+    manager: Literal["conda", "pip"]
+    platform: str
+    dependencies: Dict[str, str] = {}
+    url: str
+    hash: str
```

Comment:
Additionally i think we want to have the hash types be explicit here

Something like 

hashes: [
    {'md5': '000'},
    {'sha1': '000'},
    {'sha256': '000}'
]

That would allow us some decent future compatibility embedded into the format.

Reply:
I think this can stay as a single hash, as it refers to a single distribution in both conda and pip cases. In contrast to pip (and also poetry), the distribution is chosen for the target platform ahead of time. I suppose supporting multiple algorithms would be useful, though.